### ISSUE TITLE
[hook] `hook/generate.dart`

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#56512]
We'd like a standardized way to generate code in Dart packages that does _not_ require a user step.

* Macro's can only take Dart code as input and output.
* `package:build_runner` requires users to do a manual extra steps

In some `hook/generate.dart` this would be a `package:build_runner` equivalent, but it would be aware of the Dart/Flutter SDK and have no race conditions between saving files and running `dart` and `flutter` commands (see below).

In https://github.com/dart-lang/sdk/issues/54334, we've discussed multiple aspects w.r.t. code generators. Today we discussed some more requirements offline:

* Automatically running code generators before `pub publish` to ensure no files are outdated.
* Automatically running code generators after editing dependencies. (Or having the IDE prompt to rerun them.)
* Automatically running code generators after `pub get`, before `dart analyze`. (Only needed if the generated code is not checked in.)

Assumptions:

* The generated files are checked in. (Maybe we should aim for not checking generated files in? E.g. it would work more like macros and you'd be able to resolve to the generated files in the IDE, but they are not checked in.)
* Code generators can be heavy to run (definitely longer-running than macros).

Some open questions:

* Should it be a different hook than `hook/build.dart`? Or a mode for the build hook?
  * If it's the build hook, when do we run this hook in which modes.
  * If it's not the build hook, how do we ensure this `generate` hook is finished before running the build hook? (E.g. generate is run on-save of its dependencies, like `build_runner`. But then if you hot-restart in Flutter, you want that on-save action to be fully done first.)

Use cases:

* FFIgen, JNIgen, Swiftgen interop
* translation messages (JSON -> Dart API)

Background knowledge:

* `build.rs` enables generating of Rust files. https://doc.rust-lang.org/cargo/reference/build-script-examples.html#code-generation

Thanks @mosuem @HosseinYousefi @mkustermann @liamappelbe for the discussion! Please elaborate on things I left out.

**Comment by mosuem** [comment]
Another option would be to include `build_runner` in the SDK somehow, solving the issue with `package:build_runner requires users to do a manual extra steps`. This might run against the `macro`s efforts though.

**Comment by lrhn** [comment]
TL;DR: Either run automatically or check in the code, not both.

> Macro's can only take Dart code as input and output.

That's an assumption for the input. We don't currently let macros access the file system, but we *could*. 
If we did, then you could pass the URI of a file as an argument in the macro application, have the macro load the file (in a way that can be cached in the macro server between phases) and then it can generate code from that data too.

The alternative is to take the entire file, convert it to a Dart constant value, and have that in the macro annotation, which is just a two-step code generation where you generate the macro application code. (I'd totally do that.)

> Assumptions:
>
> * The generated files are checked in. 

That's a very strong warning signal for me. Never check in *auto*-generated code.

It's OK to have a generator that you run *manually* to generate some code, and then you check in the generated code. It's a *code updater* then, not a code generator.
When people check out the code, the code they see is the code they get, nobody will overwrite it with something else. It's stable.

It's also OK to have an automatically run generator, which generates files on demand, before they are needed.
Those files should *not* be checked in. If the code that is checked in can be changed without you asking for it, then you can't trust the repository code anyway. It's better to have no code then.


If an automatically run generator is for Dart code, macros should be *perfect* for that. That's exactly the job macros are designed for, as long as they can get the data needed to generate the code. It's automatically generated code, guaranteed up-to-date, that you can still inspect, because every tool that processes Dart programs are aware of macros. It avoids any need ever storing the code, and checking it in.

Now, for non-Dart output, macros are obviously not the answer. There will need to be some code to generate such files. And a standard for where to place them. (Should the be in `lib/` if another package depends on them?)

If files are checked in, I'd be more comfortable with a non-automatically run step, where you have to run `dart generate` to create/update the generated files. There is no need for someone just depending on a Pub packge to run `dart generate` for that package, if all generated files are checked in anyway.

> Code generators can be heavy to run (definitely longer-running than macros).

That doesn't compute for me. Macros *are* code generators. You may have some heavy use-cases in mind, but whatever framework we create here should be able to handle both small and big tasks. And so should macros.

**Comment by dcharkes** [comment]
> It's OK to have a generator that you run _manually_ to generate some code, and then you check in the generated code. It's a _code updater_ then, not a code generator.

Hm, I don't think we ever call these things _code updaters_, maybe we should. Then the hook should be called `hook/update_code.dart?` (My 2 cents, "generating" is a more understandable verb than "updating" in this context. Even if files are checked in such as with FFIgen/JNIgen/build_runner.)

If we'd like to hook `build_runner` into this, then let's look at build runners terminology:

> The `build_runner` package provides a concrete way of generating files using Dart code.

https://pub.dev/packages/build_runner

cc @davidmorgan

**Comment by davidmorgan** [comment]
Thanks Daco! Looking forward to getting into the discussion here.

Re: checking in generated code, I don't think that's quite the right distinction to make. What's important is whether, after generation, the generator _retains ownership_ of the output: has an opinion about whether it's correct.

I've been calling generators that do not retain ownership of their output "offline generators". A web page can be an offline generator, you copy its output into your source tree and you're done.

An _online_ generator is the type of generator you want to run automatically.

Whether you check in the output of online generators or not is mostly a secondary issue related to convenience of other tools, for example you might do it so the output and diffs are visible on GitHub. And for pub, you publish the generated output today because otherwise your package doesn't work :)

Re: code generators being heavy to run--I'm working on this one, they should usually be about the same cost as lints.

**Comment by HosseinYousefi** [comment]
I haven't benchmarked it but JNIgen is quite fast on my machine. Especially if we're not recompiling code. 

Both tools are not going be called frequently if we're just generating ready libraries because they don't change. One case where we will be updating is pigeon.

@tarrinneal is working on making pigeon work with FFIgen and JNIgen. We're generating the interfaces for Kotlin and Swift from a Dart IDL using `build_runner`. We could skip a parsing step for FFIgen and JNIgen since we already know what we'll be generating, so the tools will not have to parse (using clang for FFIgen and ASM for JNIgen) anymore which would improve their speed by a lot as the codegen part is not the bottleneck.

**Comment by dcharkes** [comment]
Some notes from a discussion with @mosuem: 

* The generate hook should have a `watch` mode that can integrate with the IDE/SDK. In this mode, the process should keep running, and the output should be probably a socket that can be talked to.
  * What exactly the communication is with sockets/ports should be experimented with. E.g. Does the IDE/SDK send file uri's that were changed, or file contents as well? Does the hook save generated files to disk and send back the file uri's, or does it send back the new contents of files (and the SDK then saves them to disk)?
* `batch` mode would be as already described in the previous posts.
  * The way that the `hook/generate.dart` should be written is with an abstraction that works both for `batch` and `watch` mode.
    * `build_runner`, `FfiGen`, and `JniGen` should provide a constructor that works for that.
* Code generators need to be able to communicate that they don't support `watch` mode. (Hopefully we can make it easy for custom code generator to support watch mode.)
  * The `hook/generator.dart` needs to be able to communicate that watch mode is only supported for a subset of code generators (there might be more than one code generator).
* Code generators need to be able to communicate that they might not be able to be rerun on every host.
  * For example, generating bindings against MacOS system APIs cannot be rerun on a Linux or Windows machine. But it should be possible to work on a package locally which has MacOS system API bindings and Linux system API bindings.


```dart
// Sketch for a hook/generate.dart

import 'package:hook/generate.dart';
import 'package:ffigen/ffigen.dart';
import 'package:build_runner/build_runner.dart';

void main(List<String> arguments) async {
  // This function doesn't return in `watch` mode, but does in batch mode.
  await generate(arguments, (input, output) async {
    final packageName = input.packageName;
    final ffigen1 = FfiGen(
      // ...
    );
    final ffigen2 = FfiGen(
      // ...
    );
    final buildRunner = BuildRunner(
      // ...
    );
    final customGenerator = MyGenerator(
      // ... might not support watch mode
    );
    final generators = [ffigen1, ffigen2, buildRunner, customGenerator];
    for (final generator in generators) {
      // input contains whether it's `batch` or `watch` mode.
      await generator.run(
        input: input,
        output: output,
        logger:
            Logger('')
              ..level = Level.ALL
              ..onRecord.listen((record) {
                print(record.message);
              }),
      );
    }
  });
}
```

**Comment by davidmorgan** [comment]
I suspect that delegating "watch" responsibility to each generator is going to get complicated: it means there is no overall control, each generator will independently pick up changes and produce outputs.

For cases where generators are independent this is just awkward, for cases where one generator output feeds into another's input it starts to break :)

It might actually be simpler to turn fiigen and intl into badly behaved build_runner generators that do extra work on the side, in order to get build_runner to manage running them correctly interleaved with build_runner generators.

**Comment by mosuem** [comment]
> I suspect that delegating "watch" responsibility to each generator is going to get complicated: it means there is no overall control, each generator will independently pick up changes and produce outputs.

I think the idea is to a single watcher, which notifiers each generator on changes.

> For cases where generators are independent this is just awkward, for cases where one generator output feeds into another's input it starts to break :)

I don't see the awkward part - and I would just disallow feeding inputs into other generators for starters. Long term, we might think about having a similar mechanism as for build hooks, where a DAG is constructed and invoked in order.

**Comment by davidmorgan** [comment]
> > I suspect that delegating "watch" responsibility to each generator is going to get complicated: it means there is no overall control, each generator will independently pick up changes and produce outputs.
> 
> I think the idea is to a single watcher, which notifiers each generator on changes.

Ah okay--possibly I was reading too much into the pseudocode :)

> > For cases where generators are independent this is just awkward, for cases where one generator output feeds into another's input it starts to break :)
> 
> I don't see the awkward part - and I would just disallow feeding inputs into other generators for starters. Long term, we might think about having a similar mechanism as for build hooks, where a DAG is constructed and invoked in order.

build_runner generators need to read config and need to know what files are on disk in order to figure out their inputs+outputs; so for example a generator might match `lib/foo/*bar.dart` then generate `lib/foo/*bar.g.dart`. Then another generator might match on that output and produce further output.

build_runner generators that use the analyzer need to track transitive imports and count those as inputs; and this can include files that don't exist at the start of the build, and generated files that have imports.

So ... yeah, complicated ;)

It's fine to start simple, of course.

**Comment by dcharkes** [comment]
> for cases where one generator output feeds into another's input

That's an interesting situation indeed. And if we have multiple generators inside a single `hook/generate.dart` we would need to order them inside there. And we would also need to have the watch somehow forward changes _or_ have the SDK fire a new set of file changes against the same process/socket. (We cannot rely on build_runner doing the ordering for us.)

Maybe we should have different sockets for different generators if there's more than one generator specified in `hook/generate.dart`. We have a large design space here. This was just a sketch, maybe it won't work with this sketch. We probably need a prototype to know if this would work.

> and I would just disallow feeding inputs into other generators for starters

I don't think that's a good idea, we'd want composability.

> It might actually be simpler to turn ffigen and intl into badly behaved build_runner generators that do extra work on the side, in order to get build_runner to manage running them correctly interleaved with build_runner generators.

I think `build_runner` is too restrictive.

> `build_runner` relies on determining a static build graph before starting a build

https://github.com/dart-lang/build/blob/master/docs/faq.md#why-do-builders-need-unique-outputs

In general, code generators can traverse the file system and come up with dependencies at runtime. That's why you produce a `deps` file as the output, instead of knowing all the inputs up front.

And in general, code generators can produce multiple output files, and come up with these at runtime. (Packaging outputs in a zip file to have a single output seems like a weird workaround. And we would also have to package the whole file system in a zip to have a single predefined input conceptually. Of course that would not work with caching, that zip would change all the time.)

My hope would be that `hook/generate.dart` would put less restrictions on generators. We should not be designing a system that cannot work with already existing use cases. So we need to support dynamically discovering dependencies and dynamically defining outputs. The fact that `build_runner` restricts more, and is consequently able to provide more (ordering different builders), is nice, but not applicable to the whole eco system of Dart code generators.

**Comment by davidmorgan** [comment]
This is an extremely well explored space; what `build_runner` does is based on what `bazel` does, and leads to important results about scalability and correctness.

If a generator has full freedom to define its own outputs at runtime, then in order to determine the build graph you have to _repeatedly run_ all the generators until the graph reaches a stable state. You end up with a build system that is extremely fragile and cannot be parallelized.

What `bazel` does is to require that the build graph can be computed before a generator fully runs and without any unbounded computation; and `build_runner` does something similar. This makes the build graph computation guaranteed to run quickly and without churn. Then you can get it over with and get on to the actual build :) knowing enough to parallelize.

I 100% agree with addressing the use cases of generators that want more flexible output. Archives seems to me--at first glance--to be the right solution. There is no reason for archives to break caching: quite the reverse, an archive can provide digests for itself and for the files it contains, making caching simpler. There is no reason to use archives for input, individual file inputs are fine, any input can be generated as long as it comes from a generator that can cheaply compute its outputs.

I'm sure we can come up with a simple+correct hack that works before archives are available / before we can build the needed support in build_runner. For example just running the noncompliant generators in the generate hook after build_runner in a fixed order, and forbidding any dependency in the other direction, is probably sufficient; we just hide the non-build_runner output completely from build_runner.

Thanks.

**Comment by dcharkes** [comment]
Some musings with @davidmorgan:

* Code generators are inherently ordered at package boundaries by the package import graph.
  *  `build_runner` generators can read from transitive deps, and can only output in the current package.
  * `hooks/build.dart` have the same restriction. Only read from your own package and (_direct_) deps, and only output in the current package output directory.
  * So multiple `hook/generator.dart`s in a workspace would be ordered by the package graph.
* Within a `hook/generate.dart` it would be nice if we could more fine-grained ordering/parallelization of generators.
  * This requires knowing the inputs and outputs up front.
    * The way that `bazel` forces this is by basically doing globs over the inputs in the "planning phase". This is an overapproximation, but suffices. (For Dart compilation its all the Dart files in a package and all the other build rules which could be dart packages with their Dart files again.)
      * We can probably do something similar for FFIgen and JNIgen doing globs over the include directories in a "planning-phase" and by such means wrap FFIgen and JNIgen inside build runner.
  * To use `build_runner` to order multiple generators, and to have generators such as FFIgen, JNIgen, intl, win32, we most likely need to extend `build_runner`s capabilities slightly.
    * For example: Inputs outside a Dart package (system headers)
    * We will explore trying to wrap FFIgen, JNIgen, win32 in `build_runner` to see where capabilities need to be extended.

**Comment by gaaclarke** [comment]
I think this could handle pigeon as well.  When I've discussed this with others in the past they had concerns about increasing build times.  The point I made is that buildtime is no different than runtime and we trust users to manage that situation.  I think it's important that we have a way to audit build time execution though so that users can evaluate the value they are getting versus the added time to builds.  Unity3D had a good tool for this from what I recall (though it was more geared toward asset packaging time?).

Something like:
```
building foo:

executing generate.dart:
--- foo_package: 101ms
--- bar_package: 34ms
...
--- total: 1010ms
```

It may even be worth communicating dependency graphs too to help track it down.  So if the project depends on `foo` and `foo` depends on `bar` and `bar` has the `generate.dart` it would be `--- foo/bar: Xms`.

**Comment by HosseinYousefi** [comment]
One use case I want to cover using generate hooks is CI and the discoverability of generator tools. Currently, when a project uses a codegen tool, it normally exists as one of potentially many dev-dependencies. There is no obvious way to check if all the codegen tools have been run with the latest versions. Each tool can have its own way of checking this or simply no way at all, and it's up to each package author to regenerate each codegen and diff to see if any files in the project changed.

This is in contrast to tools like `dart format`, where we have a `--set-exit-if-changed` flag that returns 1 if the tool changes any code. With a unified generation hook that passes a generation config to all codegen tools which includes this flag, a CI check will be as simple as running `dart generate --set-exit-if-changed`.

Another thing that this workflow simplifies is the discoverability of such tools. A new contributor to a Dart/Flutter project doesn't have to learn about a bunch of scripts that are different from project to project; instead, running `dart generate` simply runs all of the codegen tools available in the project.

Having the generation hook inside `hook/generate.dart` limits it to a single package. However, we probably want the generation hook to be per workspace. Where should `generate.dart` reside in this case?

We also discussed having a `dart generate --watch` with @davidmorgan, @dcharkes, and @goderbauer. This is analogous to `build_runner`'s watch mode, but it seems like that problem space is much vaster, so for now, we don't want to tackle it.

**Comment by liamappelbe** [comment]
> Having the generation hook inside `hook/generate.dart` limits it to a single package. However, we probably want the generation hook to be per workspace. Where should `generate.dart` reside in this case?

The workspace is still a package, so it could be in the workspace's `hook` directory. Or we could say that running generate in the root of a workspace runs it in all the workspace's nested packages. I'd prefer the latter since it means that the generation scripts can remain modular.

**Comment by dcharkes** [comment]
> > Having the generation hook inside `hook/generate.dart` limits it to a single package. However, we probably want the generation hook to be per workspace. Where should `generate.dart` reside in this case?
> 
> The workspace is still a package, so it could be in the workspace's `hook` directory. Or we could say that running generate in the root of a workspace runs it in all the workspace's nested packages. I'd prefer the latter since it means that the generation scripts can remain modular.

I'm leaning towards per package indeed. And, probably in the order of dependencies, the same order as build hooks. That way generators can be taking each others' outputs as inputs again. And, maybe control workspace vs single package based on working dir.

Workspaces create an interesting question about generators that generate from one package into another. For example a single config that generates into 3 packages. Currently, `build_runner` does not allow this, but handwritten generators can (and are) doing this. If one would want correct ordering in that case, the generator package needs to be a dev dependency of the packages it's generating in to. (However, maybe we should be telling people that they should move all the logic to be accessible from `lib/` or `bin/` and generate in each packages' generate hook separately.)

**Comment by dcharkes** [comment]
> Workspaces create an interesting question about generators that generate from one package into another.

I've spent some more time looking into this with the various code generators we're using on `dart-lang/native`. I think the following approach would work:

* No generate hook and no generators at the workspace level, only generators inside packages.
  * Rationale: Even if a generator would work with the `packages_config.json` that is generated, one would not be able call the generator from another Dart package due to not being able to import it. And I believe it's common to have generators both being used from a CLI as well as from a Dart API.
* Generators may generate code into their own package, generators may generate code into packages depending on them (via dev dependencies), generators may not generate code into their dependencies.
  * Rationale: Running the generate hooks in topological ordering of dependencies ensures that no generate hook will have their inputs change after being run.
  * Generating into packages that depend on the generator may be avoided by letting the generator package expose a Dart API and adding a generate hook to every individual package.
  * However, there are cases where it is simply easier in the context of a workspace to write the generator code as a single script to avoid having to reparse packages.

If you have a generator that is used throughout the workspace my suggestion is to create a `workspace_generators` package and add that as a dev dependency to all the packages in the workspace that it generates in to.

**Comment by goderbauer** [comment]
> Generators may generate code into their own package, generators may generate code into packages depending on them (via dev dependencies), generators may not generate code into their dependencies.

I assume this is limited to packages within the same workspace or packages that otherwise know about each other? You are not  envisioning a generic mechanism where any package can declare a dev dependency on a generator to have that generator generate something into that package without the generator knowing anything about that package?

> However, there are cases where it is simply easier in the context of a workspace to write the generator code as a single script to avoid having to reparse packages.

Wondering what this would mean for usability/discoverability, e.g. for someone who is new to a project. 
* What happens if I run `dart generate` in the root of the workspace? Would this run the generator hooks of all packages in the workspace?
* What if I run `dart generate` in a package into which another package generators something? Will this rerun that generator? If no, how would I discover where I need to run the command to update the generated output? If yes, how would we identify the generators that actually affect this package and rerun those?

From a usability standpoint I think `dart generate` in the workspace root should run the generators of all packages and running `dart generate` inside a package should regenerate whatever is inside that packages. Otherwise, we are sort of back to square one where you don't know what to run where to regenerate your stuff after changes.

> to avoid having to reparse packages.

If we were to disallow cross-package generation, how expensive is it actually to reparse? I'd assume you could reuse the code for this across packages by putting it in a helper package. Are you concerned that it would be too slow?

**Comment by dcharkes** [comment]
> > Generators may generate code into their own package, generators may generate code into packages depending on them (via dev dependencies), generators may not generate code into their dependencies.
> 
> I assume this is limited to packages within the same workspace or packages that otherwise know about each other? You are not envisioning a generic mechanism where any package can declare a dev dependency on a generator to have that generator generate something into that package without the generator knowing anything about that package?

No, this would have to be in a workspace.

> > However, there are cases where it is simply easier in the context of a workspace to write the generator code as a single script to avoid having to reparse packages.
> 
> Wondering what this would mean for usability/discoverability, e.g. for someone who is new to a project.
> 
> * What happens if I run `dart generate` in the root of the workspace? Would this run the generator hooks of all packages in the workspace?

Yes.

> * What if I run `dart generate` in a package into which another package generators something? Will this rerun that generator?

No, a generator generating into the current package should not be reading from the current package. So if you only work on that package, no generators generating into that package need to be re-run. Note that you cannot have a "formatter" sitting in the generate hook of another package, because it will need to read the current package. A formatter needs to be configured in the generate hook of the current package.

> If no, how would I discover where I need to run the command to update the generated output?

I think it should be best practice that _every_ generated file contains a header containing the command on how to regenerate it. (I haven't come up with a way to enforce that via the generate hook. If we would like to enforce something like that we would need to come up with a Dart API that returns strings and file path tuples and add it inside that Dart API. But that's rather fragile if you also have to deal with copyright headers being first and non-Dart files being generated.)

> From a usability standpoint I think `dart generate` in the workspace root should run the generators of all packages and running `dart generate` inside a package should regenerate whatever is inside that packages. Otherwise, we are sort of back to square one where you don't know what to run where to regenerate your stuff after changes.

➕ 

> > to avoid having to reparse packages.
> 
> If we were to disallow cross-package generation, how expensive is it actually to reparse? I'd assume you could reuse the code for this across packages by putting it in a helper package. Are you concerned that it would be too slow?

If generated package in one package depends on generated code in another package, and both are generated by the same generator, there are two options:

1. Having the generators generate a `symbols.json` file that's read during the generation of the second package. This is developer overhead: You need to come up with a format to capture the information. (FFIgen, JNIgen)
2. Or reparsing + reading the configuration file from the first package in the second package to know what was already generated in the first package and doesn't have to be regenerated in the second package. If the configuration in the first package is Dart API calls, and you don't want this to be part of the public API, that means you need to do a `lib/src/` import of package 1 in package 2. (`json_syntax_generator` for `hooks`, `code_assets` and `data_assets`.)

(Option 1 allows for generated code depending on generated code in another package outside the workspace. Option 2 does not, you should not have a src/ import even in generators if the package is resolved via pub and the code is only constrained by semantic versioning.)

Yeah, maybe we should simply disallow cross-package generation. It simplifies the mental model.

**Comment by scheglov** [comment]
> No, a generator generating into the current package should not be reading from the current package.

Isn't it the most useful case: read some files from a package to generate some other files?
E.g. in the analyzer: read `ast.dart` and generate `visitor.g.dart` for all AST nodes.

**Comment by bwilkerson** [comment]
Sorry, I know I'm late to the party, but I have an answer to a question asked earlier on.

> - The generate hook should have a watch mode that can integrate with the IDE/SDK. In this mode, the process should keep running, and the output should be probably a socket that can be talked to.

To the best of my knowledge, the IDEs don't support allowing a separate process to make a socket connection with them (probably for security reasons). The language servers they run are all started by the IDE (or more accurately an IDE plugin) and communicate with it over stdin/stdout.

>   - What exactly the communication is with sockets/ports should be experimented with. E.g. Does the IDE/SDK send file uri's that were changed, or file contents as well? Does the hook save generated files to disk and send back the file uri's, or does it send back the new contents of files (and the SDK then saves them to disk)?

The IDE/SDK sends the updated file content as the user is typing. The hook has to send back the new contents of the files. The IDE can reject the changes if the files being updated have changed outside the hooks. If the IDE accepts the changes, then whether they are saved to disk or just used to update an edit buffer is up to the IDE, usually based on user preferences.

**Comment by dcharkes** [comment]
> > No, a generator generating into the current package should not be reading from the current package.
> 
> Isn't it the most useful case: read some files from a package to generate some other files? E.g. in the analyzer: read `ast.dart` and generate `visitor.g.dart` for all AST nodes.

It's totally fine to read and write files from the current package with a generator in the current package. (I should have phrased the above sentence as "No, a generator inside a package the current package is depending on generating into the current package should not be reading from the current package.")

**Comment by flar** [comment]
I have a use case that would aim to make interacting with fragment shader programs from Flutter code easier. Fragment shaders have "uniform values" in them that you set to configure their output. Currently in order to do that you have to correlate the locations of the uniform variable declarations within the fragment program with "indices" to set the values. The engine team has made this less painstaking by allowing a program to look up the indices by name at run time, but the result doesn't really capture the concept that the shader is an object that has fields to set.

I'm working on a package that has a script to scan the project for shaders (from the data in pubspec.yaml) and to automatically generate some Dart code that emulates the "set shader value by index" procedure as if it were a field in a Dart object, but it requires a manual step as indicated above. It would be nice to be able to automate "developer changes the .frag file listed in pubspec.yaml and the build process automatically regenerates the generated Dart wrappers". The conversion script works very well and produces objects that make the instances of the fragment program appear to have settable fields, except for the need to run it manually.

**Comment by ganeshrvel** [comment]
@dcharkes

One additional requirement I'd like to suggest is that the generate hook should have access to environment variables and other build-time inputs. Also, if the generate hook fails (either by throwing an exception or otherwise), the build should fail immediately and not proceed with analysis or compilation to avoid building with missing or partially generated source files.

**Comment by davidmorgan** [comment]
A few updates

 - build_runner now has `--only-check` to check if generated files are correct without changing anything
 - build_runner is a lot faster now :)
